src/robot/libraries/Telnet.py

```diff
@@ -491,8 +531,17 @@ def __init__(self, host=None, port=23, timeout=3.0, newline='CRLF',
                  encoding='UTF-8', encoding_errors='ignore',
                  default_log_level='INFO', window_size=None, environ_user=None,
                  terminal_emulation=False, terminal_type=None,
-                 telnetlib_log_level='TRACE'):
-        telnetlib.Telnet.__init__(self, host, int(port) if port else 23)
+                 telnetlib_log_level='TRACE', connection_timeout=None):
+        if connection_timeout is None:
+            telnetlib.Telnet.__init__(self, host, int(port) if port else 23)
+        else:
+            try:
+                telnetlib.Telnet.__init__(self, host,
+                                          int(port) if port else 23,
+                                          connection_timeout)
+            except socket.timeout:
+                raise AssertionError("Connection timeout after %s."
+                                     % secs_to_timestr(connection_timeout))
```

Comment:
Some users do use the TelnetLibrary programmatically. Raising different exception types (socket.timeout, AssertionError) based on what timeout was used is a bit unclean. Also on my linux machine I get socket.error when the system refuses the connection and this fails the test "Set Connection Timeout Keyword".

If socket.timeout is caught then we should catch also it in the case where the timeout occurs with default timeout or then alternatively, we should rise the same exeption (socket.timeout). I discussed this with @pekkaklarck and we were thinking that maybe easiest would be not to catch any exceptions here and just let the original socket exception to go through.

I assume that in your environment that test passes, so presumably you get socket.timeout instead of socket.error. Does you network have this machine "172.16.254.254" or are you running on some other platform than I do? (Linux mint, python 2.7.3)